packages/contractkit/src/utils/web3-utils.ts

```diff
@@ -71,3 +71,68 @@ class WrappingSubprovider extends Subprovider {
     })
   }
 }
+
+// tslint:disable-next-line:max-classes-per-file
+export class CeloProvider implements Provider {
+  private static createNewProviderWithLocalAccount(existingProvider: Provider): Web3ProviderEngine {
+    // Create a Web3 Provider Engine
+    const providerEngine = new Web3ProviderEngine()
+    // Use the existing provider to route all other requests
+    const wrappingSubprovider = new WrappingSubprovider(existingProvider)
+    debug('Setting up providers...')
+    providerEngine.addProvider(wrappingSubprovider)
+    return providerEngine
+  }
+
+  private web3ProviderEngine: Web3ProviderEngine
+  private signingProviders: CeloPrivateKeyWalletProvider[]
+  private signingAddresses: string[]
+
+  constructor(readonly existingProvider: Provider) {
+    this.web3ProviderEngine = CeloProvider.createNewProviderWithLocalAccount(existingProvider)
+    this.signingProviders = []
+    this.signingAddresses = []
+  }
+
+  addAccount(privateKey: string): CeloProvider {
+    if (this.signingAddresses.includes(privateKey)) {
+      debug('Account is already added')
+      return this
+    }
+
+    const localAccountProvider = new CeloPrivateKeyWalletProvider(privateKey)
+    // When a private key has already been added, add a new one before
+    // all other providers.
+    this.signingProviders.splice(0, 0, localAccountProvider)
+    // I could not find a better way to do this, so, I had to
+    // access the private `_providers` field of providerEngine
+    // @ts-ignore-next-line
+    this.web3ProviderEngine._providers.splice(0, 0, localAccountProvider)
+    localAccountProvider.setEngine(this.web3ProviderEngine)
+    this.signingAddresses.push(localAccountProvider.getAccount())
+    return this
+  }
+
+  on(event: string, handler: () => void): void {
```

Comment:
we should be smarter here.
IPC and WS provider have `on`, and thus "support subscriptions" in web3js terms, while HTTP doesn't. If we set this provider, web3js will understand that WE support subscriptions always

Reply:
solved